README.rst
Kinto signer
############

|travis| |coveralls|

.. |travis| image:: https://img.shields.io/github/workflow/status/Kinto/kinto-signer/test/master
    :target: https://github.com/Kinto/kinto-signer/actions

.. |coveralls| image:: https://coveralls.io/repos/github/Kinto/kinto-signer/badge.svg?branch=master
    :target: https://coveralls.io/github/Kinto/kinto-signer?branch=master

**Kinto signer** is a `Kinto <https://kinto.readthedocs.io>`_ plugin
that signs records with a `content signature <https://github.com/mozilla-services/autograph/blob/3dc9cfc/signer/contentsignature/README.rst>`_
to guarantee their integrity and authenticity.


How does it work?
=================

**Kinto signer** uses two collections:

* The *source*, where the authors create/update/delete records.
* The *destination*, where the clients obtain the records and their signature.

When the *source* collection metadata ``status`` is set to ``"to-sign"``, it will:

#. grab the whole list of records in this *source* collection
#. update the *destination* collection records with the recent changes
#. serialize the result in a Canonical JSON form (*see below*)
#. compute a signature using the configured backend
#. update the *destination* collection metadata ``signature`` with the information
   obtain form the signature backend
#. set the *source* metadata ``status`` to ``"signed"``.

A publishing workflow can be enabled (see below).

.. warning::

    The current implementation assumes the destination collection will be
    readable anonymously and won't be writable by anyone.
    (See `Kinto/kinto-signer#55 <https://github.com/Kinto/kinto-signer/issues/55>`_)


Content-Signature protocol
--------------------------

Kinto-signer produces signatures for the content of Kinto collections using
`ECDSA <https://fr.wikipedia.org/wiki/Elliptic_curve_digital_signature_algorithm>`_
with the P-384 strength.

* ``content = {"data": sorted(records, key=operator.itemgetter("id")), "last_modified": f"{last_modified}"}``
* The content is prepended with ``Content-Signature:\x00`` prior to signing.
* The signature is produced with ECDSA on P-384 using SHA-384.
* The signature is returned as encoded using URL-safe variant of base-64.

References:

* `Content Signature <https://github.com/mozilla-services/autograph/blob/e7c33d6/signer/contentsignature/README.rst>`_ on Mozilla Autograph
* `canonical json implementation <https://github.com/mozilla-services/python-canonicaljson-rs>`_
* Signature validation in Firefox is done using the `Personal Security Manager <https://developer.mozilla.org/en/docs/Mozilla/Projects/PSM>`_.


Setup
=====

To install this plugin in a Kinto server, a few configuration variables need
to be set.

Here is an example of what a configuration could look like:

.. code-block:: ini

  kinto.includes = kinto_signer

  kinto.signer.resources =
      /buckets/source                         -> /buckets/destination
      /buckets/source/collections/collection1 -> /buckets/destination/collections/collection2
      /buckets/bid/collections/cid            -> /buckets/bid/collections/cid2

+---------------------------------+--------------------------------------------------------------------------+
| Setting name                    | What does it do?                                                         |
+=================================+==========================================================================+
| kinto.signer.resources          | The source URIs (bucket or collection) on which signatures should be     |
|                                 | triggered and the destination where the data and the signatures will     |
|                                 | end-up.                                                                  |
|                                 |                                                                          |
|                                 | In the case buckets URIs are specified, every collection in the source   |
|                                 | bucket will be reviewed/signed, review and destination will keep the     |
|                                 | same id.                                                                 |
+---------------------------------+--------------------------------------------------------------------------+
| kinto.signer.signer_backend     | The python dotted location to the signer to use. By default, a local     |
|                                 | ECDSA signer will be used. Choices are either                            |
|                                 | ``kinto.signer.signer.local_ecdsa`` or ``kinto.signer.signer.autograph`` |
|                                 | Have a look at the sections below for more information.                  |
+---------------------------------+--------------------------------------------------------------------------+
| kinto.signer.allow_floats       | Allow float values in records (default: ``False``).                      |
|                                 | Toggling this setting to ``True`` can lead to signature verification     |
|                                 | errors in clients.                                                       |
|                                 | See ``kinto_signer.listeners.prevent_float_value``                       |
+---------------------------------+--------------------------------------------------------------------------+

Configuration for the (default) ECDSA local signer
--------------------------------------------------

+---------------------------------+--------------------------------------------------------------------------+
| Setting name                    | What does it do?                                                         |
+=================================+==========================================================================+
| kinto.signer.ecdsa.private_key  | Absolute path to the ECDSA private key to use to apply the signatures    |
+---------------------------------+--------------------------------------------------------------------------+
| kinto.signer.ecdsa.public_key   | Absolute path to the ECDSA private key to use to verify the signature    |
|                                 | (useful if you just want to use the signer as a verifier)                |
+---------------------------------+--------------------------------------------------------------------------+


Configuration for the Autograph signer
--------------------------------------

Kinto signer can integrate with the
`Autograph <https://github.com/mozilla-services/autograph>`_ server version 2.
To do so, use the following settings:

+------------------------------------+--------------------------------------------------------------------------+
| Setting name                       | What does it do?                                                         |
+====================================+==========================================================================+
| kinto.signer.autograph.server_url  | The autograph server URL                                                 |
+------------------------------------+--------------------------------------------------------------------------+
| kinto.signer.autograph.hawk_id     | The hawk identifier used to issue the requests.                          |
+------------------------------------+--------------------------------------------------------------------------+
| kinto.signer.autograph.hawk_secret | The hawk secret used to issue the requests.                              |
+------------------------------------+--------------------------------------------------------------------------+


Workflows
---------

A workflow can be enabled on the source collection ``status``.

The workflow is basically ``work-in-progress`` → ``to-review`` → ``to-sign`` → ``signed`` and
makes sure that:

* the collection is reviewed before being signed
* the user asking for review is the not the one approving the review
* the user asking for review belongs to a group ``editors`` and
  the one approving the review belongs to ``reviewers``.

+----------------------------------+---------------+--------------------------------------------------------------------------+
| Setting name                     | Default       | What does it do?                                                         |
+==================================+===============+==========================================================================+
| kinto.signer.to_review_enabled   | ``false``     | If ``true``, the collection ``status`` must be set to ``to-review`` by a |
|                                  |               | different user before being set to ``to-sign``.                          |
+----------------------------------+---------------+--------------------------------------------------------------------------+
| kinto.signer.group_check_enabled | ``false``     | If ``true``, the user setting to ``to-review`` must belong to the        |
|                                  |               | ``editors`` group in the source bucket, and the one setting to           |
|                                  |               | ``to-sign`` must belong to ``reviewers``.                                |
+----------------------------------+---------------+--------------------------------------------------------------------------+
| kinto.signer.editors_group       | ``editors``   | The group id that is required for changing status to ``to-review``       |
+----------------------------------+---------------+--------------------------------------------------------------------------+
| kinto.signer.reviewers_group     | ``reviewers`` | The group id that is required for changing status to ``to-sign``         |
+----------------------------------+---------------+--------------------------------------------------------------------------+

.. warning::

    The ``editors`` and ``reviewers`` groups are defined in the **source bucket**
    (e.g. ``/buckets/staging/groups/editors``).

    The ``editors`` and ``reviewers`` groups can have placeholders that are resolved
    with the source **source bucket/collection**
    (e.g. ``group:/buckets/{bucket_id}/groups/{collection_id}-reviewers``).

See `Kinto groups API <http://kinto.readthedocs.io/en/stable/api/1.x/groups.html>`_ for more details about how to define groups.

The above settings can be set or overriden by bucket using the ``<bucket_id>_`` prefix or by collection using the ``<bucket_id>_<collection_id>_`` prefix.
For example:

.. code-block:: ini

    kinto.signer.staging.group_check_enabled = true
    kinto.signer.staging.to_review_enabled = true
    kinto.signer.staging.certificates.group_check_enabled = false
    kinto.signer.staging.certificates.to_review_enabled = false
    kinto.signer.staging.certificates.reviewers_group = certificates-reviewers

If the review process is enabled, it is possible to configure a *preview*
collection, that will be updated and signed when the status is set to ``to-review``.
This *preview* collection can be used by clients to test and validate the changes
before approving them.

If a resources entry contains a semi-column separated **triplet**, then a preview
collection will be enabled.

.. code-block:: ini

  kinto.signer.resources =
      /buckets/staging            -> /buckets/preview            -> /buckets/blog
      /buckets/bid/collections/c1 -> /buckets/bid/collections/c2 -> /buckets/bid/collections/c3


.. image:: workflow.png


The editors and reviewers groups are automatically created when the source collection is created.


Multiple certificates
---------------------

Using above settings, every collections is signed with the same key.
But it is also possible to define multiple signers, per bucket or per collection.

Settings can be prefixed with bucket id:

.. code-block:: ini

    kinto.signer.signer_backend = kinto_signer.signer.autograph
    kinto.signer.autograph.server_url = http://172.11.20.1:8888

    kinto.signer.<bucket-id>.autograph.hawk_id = bob
    kinto.signer.<bucket-id>.autograph.hawk_secret = a-secret


Or prefixed with bucket and collection:

.. code-block:: ini

    kinto.signer.<bucket-id>.<collection-id>.signer_backend = kinto_signer.signer.local_ecdsa
    kinto.signer.<bucket-id>.<collection-id>.ecdsa.private_key = /path/to/private.pem
    kinto.signer.<bucket-id>.<collection-id>.ecdsa.public_key = /path/to/public.pem


Usage
=====

Suppose we defined the following resources in the configuration:

.. code-block:: ini

    kinto.signer.resources = /buckets/source -> /buckets/destination

First, if necessary, we create the appropriate Kinto objects, for example, with ``httpie``:

.. code-block:: bash

    $ http PUT http://0.0.0.0:8888/v1/buckets/source --auth user:pass
    $ http PUT http://0.0.0.0:8888/v1/buckets/source/collections/collection1 --auth user:pass
    $ http PUT http://0.0.0.0:8888/v1/buckets/destination --auth user:pass
    $ http PUT http://0.0.0.0:8888/v1/buckets/destination/collections/collection1 --auth user:pass

Create some records in the *source* collection.

.. code-block:: bash

    $ echo '{"data": {"article": "title 1"}}' | http POST http://0.0.0.0:8888/v1/buckets/source/collections/collection1/records --auth user:pass
    $ echo '{"data": {"article": "title 2"}}' | http POST http://0.0.0.0:8888/v1/buckets/source/collections/collection1/records --auth user:pass


Trigger a signature operation, set the ``status`` field on the *source* collection metadata to ``"to-sign"``.

.. code-block:: bash

    echo '{"data": {"status": "to-sign"}}' | http PATCH http://0.0.0.0:8888/v1/buckets/source/collections/collection1 --auth user:pass

The *destination* collection should now contain the new records:

.. code-block:: bash

    $ http GET http://0.0.0.0:8888/v1/buckets/destination/collections/collection1/records --auth user:pass

.. code-block:: javascript

    {
        "data": [
            {
                "article": "title 2",
                "id": "a45c74a4-18c9-4bc2-bf0c-29d96badb9e6",
                "last_modified": 1460558489816
            },
            {
                "article": "title 1",
                "id": "f056f42b-3792-49f3-841d-0f637c7c6683",
                "last_modified": 1460558483981
            }
        ]
    }

The *destination* collection metadata now contains the signature:

.. code-block:: bash

   $ http GET http://0.0.0.0:8888/v1/buckets/destination/collections/collection1 --auth user:pass

.. code-block:: javascript

   {
       "data": {
           "id": "collection1",
           "last_modified": 1460558496510,
           "signature": {
               "mode": "p384ecdsa",
               "x5u": "https://bucket.example.net/appkey1.pem",
               "signature": "Nv-EJ1D0fanElBGP4ZZmV6zu_b4DuCP3H7xawlLrcR7to3aKzqfZknVXOi94G_w8-wdKlysVWmhuDMqJqPcJV7ZudbhypJpj7kllWdPvMRZkoWXSfYLaoLMc8VQEqZcb"
           }
       },
       "permissions": {
           "read": [
               "system.Everyone"
           ]
       }
   }

Tracking fields
---------------

During the review process, the *source* collection metadata will receive the following read-only fields:

- ``last_edit_by``: last user to perform change on records in the source collection
- ``last_edit_date``: date of the last records change
- ``last_review_request_by``: last user to request a review
- ``last_review_request_date``: date of the last review request
- ``last_review_by``: last user to approve a review
- ``last_review_date``: date of the last review approval
- ``last_signature_by``: last user to trigger a signature
- ``last_signature_date``: date of the last signature

.. note:

  ``last_signed_by`` can be different from ``last_review_by`` when the signature is refreshed
  (ie. status changed directly from ``signed`` to ``to-sign``).


Rollback changes
----------------

In order to reset the source (and preview) collection with the content of the destination collection (ie. last approved content), set the source to ``to-rollback``.

.. code-block:: bash

    echo '{"data": {"status": "to-rollback"}}' | http PATCH http://0.0.0.0:8888/v1/buckets/source/collections/collection1 --auth user:pass


Refresh signature
-----------------

In order to refresh the signature, set the source to ``to-resign``, the content signature
metadata will be recomputed and updated and the status restore to its previous value
(eg. ``signed`` or ``to-review``...).

This is useful when the signer certificates are rotated etc.

.. code-block:: bash

    echo '{"data": {"status": "to-resign"}}' | http PATCH http://0.0.0.0:8888/v1/buckets/source/collections/collection1 --auth user:pass


Events
======

Pyramid events are sent for each review step of the validation workflow.

Events have the following attributes:

* ``request``: current Pyramid request object
* ``payload``: same as ``kinto.core.events.ResourceChanged``
* ``impacted_records``: same as ``kinto.core.events.ResourceChanged``
* ``resource``: dict with details about source, preview and destination collection
                (as in capability).
* ``original_event``: original ``ResourceChanged`` event that was caught to
                      detect step change in review workflow.

The following events are thrown:

* ``kinto_signer.events.ReviewRequested``
* ``kinto_signer.events.ReviewRejected``
* ``kinto_signer.events.ReviewApproved``
* ``kinto_signer.events.ReviewCanceled`` (when source is rolledback)

.. important::

    The events are sent within the request's transaction. In other words, any
    database change that occurs in subscribers will be committed or rolledback
    depending of the overall response status.


Validating the signature
========================

With `kinto.js <https://github.com/Kinto/kinto.js/>`_, it is possible to define
incoming hooks that are executed when the data is retrieved from the server.

.. code-block:: javascript

    const kinto = new Kinto({
      remote: "https://mykinto.com/v1",
      bucket: "a-bucket"
    });
    const collection = kinto.collection("a-collection", {
      hooks: {
        "incoming-changes": [validateCollectionSignature]
      }});

.. code-block:: javascript

    function validateCollectionSignature(payload, collection) {
      // 1 - Fetch signature from collection endpoint
      // 2 - Fetch public key certificate
      // 3 - Merge incoming changes with local records
      // 4 - Serialize as canonical JSON
      // 5 - Verify the signature against the content with the public key
      // 6 - Return `payload` if valid, throw error otherwise.
    }

The content of the ``demo/`` folder implements the signature verification with
kinto.js and the WebCrypto API. It is `published online <https://kinto.github.io/kinto-signer/>`_
but relies on a semi-public server instance.

See also `the complete integration within Firefox <https://bugzilla.mozilla.org/show_bug.cgi?id=1263602>`_
using the `Network Security Services <https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Overview>`_.


Generating a keypair
====================

To generate a new keypair, you can use the following command::

  $ python -m kinto_signer.generate_keypair private.pem public.pem


Running the tests
=================

In order to contribute and run the full functional test suite locally you need
to have the Go language executables (e.g. `sudo apt-get install golang`)
and a ``testdb`` PostgreSQL database like for the Kinto server.

The rest of installation and setup process is taken care of automatically.

To run the unit tests::

  $ make tests

For the functional tests, run these two services in separate terminals:

::

  $ make run-kinto

::

  $ make run-autograph

And start the test suite::

  $ make functional


CHANGELOG.rst
Changelog
=========

This document describes changes between each past release.

8.1.0 (unreleased)
------------------

**Breaking changes**

- Removed ability to resign using ``to-sign`` twice. Use to ``to-resign`` instead.

**New feature**

- Copy ``schema`` field to destination metadata (fixes #518)

**Bug fixes**

- Reset the editor/reviewer comments when not specified.


8.0.1 (2021-02-23)
------------------

**Bug fixes**

- Fix behaviour with preview collection when review is disabled (#1327)


8.0.0 (2020-11-19)
------------------

**New feature**

- Add comments to sign-off events attributes

**Breaking changes**

- Replaced custom Canonical JSON with ``python-canonicaljson-rs``
- Removed CloudFront invalidation code (#1256)


7.0.0 (2020-09-11)
------------------

**Breaking changes**

- Records cannot contain float values anymore. See ``signer.allow_floats`` settings to toggle old behaviour.


6.1.0 (2020-03-27)
------------------

- Send resource events on rollback (#894)


6.0.2 (2020-01-07)
------------------

- Do not recreate tombstones on rollback (fixes #832)


6.0.1 (2019-12-05)
------------------

- Fix records dropped since #780 (fixes #816)


6.0.0 (2019-11-15)
------------------

**Breaking changes**

- Records timestamps are now bumped when copied to preview/destination collections (fixes #779)


5.2.1 (2019-10-03)
------------------

- Fix crash in collection deletion safety check (fixes #709)


5.2.0 (2019-09-26)
------------------

**New features**

- Log Autograph requests references (#687)
- Set status of new collections to signed (fixes #683)
- Fix collection deletion safety check (fixes #684)


5.1.0 (2019-08-27)
------------------

**New Features**

- Add StatsD counter for number of changes approved (fixes #606)
- Prevent deletion of preview or destination if in use (fixes #641)
- Add ability to rollback changes (fixes #260)

**Internal changes**

- Use black to autoformat the code base. (#617)
- Use f-string to format strings. (#619)


5.0.1 (2019-07-26)
------------------

**Bug Fixes**

- Make sure the DB transaction is rolledback when signing fails (fixes #591)


5.0.0 (2019-04-04)
------------------

**Breaking changes**

- Do not invalidate CloudFront on signature refresh (fixes #430)


4.0.1 (2019-01-30)
------------------

**Security issue**

- Signer parameters were displayed in capabilities. Fixed in #326.


4.0.0 (2019-01-22)
------------------

**Bug fixes**

- Fix inconsistencies when source records are deleted via the DELETE /records endpoint (fixes #287)

**Breaking changes**

- Require Kinto >= 12.0.0


3.3.9 (2018-12-10)
------------------

**Internal Changes**

- Do not use the count value from ``storage.get_all()``


3.3.8 (2018-11-27)
------------------

- Fix "RuntimeError: OrderedDict mutated during iteration" (#283).


3.3.7 (2018-11-20)
------------------

**Bug fixes**

- If ``to_review_enabled`` is False, the preview collection is not created, nor updated (fixes #279)
- Show collections with specific settings in capabilities


3.3.6 (2018-11-08)
------------------

**Bug fixes**

- Fix Canonical JSON serialization of zero
- Allow installing ``kinto-signer`` with ``--no-deps`` in order to import ``kinto_signer.serializer.canonical_json()`` without the Pyramid ecosystem


3.3.5 (2018-11-06)
------------------

**Bug fixes**

- Fix Canonical JSON about float numbers to conform with `ECMAScript V6 notation <https://www.ecma-international.org/ecma-262/6.0/#sec-tostring-applied-to-the-number-type>`_


3.3.4 (2018-10-25)
------------------

**Bug fixes**

- Prevent events to be sent if status is not changed (#268)

**Internal Changes**

- Rewrite e2e to use ``.get_records_timestamp()`` (#258)
- Enable Wheel distribution (fixes #271)


3.3.3 (2018-10-10)
------------------

- Allow refresh of signature even if the collection was never signed (#267)


3.3.2 (2018-08-20)
------------------

- Support kinto 10.0.0, which allowed some simplifications (#264).


3.3.1 (2018-08-17)
------------------

- Failed artifact produced by mistake. Please ignore.


3.3.0 (2018-07-24)
------------------

**New features**

- Allow to refresh the signature when the collection has pending changes (fixes #245)


3.2.5 (2018-07-05)
------------------

**Bug fixes**

- Fix ``scripts/e2e.py`` script to work with per-bucket configuration
- Prevent kinto-attachment to raise errors when attachments are updated (fixes #256)i

3.2.4 (2018-05-30)
------------------

**Bug fixes**

- Fix CloudFront invalidation request with multiple paths (fixes #253)


3.2.3 (2018-05-07)
------------------

**Bug fixes**

- Fix crash on collection delete (fixes #248)


3.2.2 (2018-05-02)
------------------

**Bug fixes**

- Cleanup preview and destination when source collection is deleted (fixes #114)


3.2.1 (2018-04-25)
------------------

**Bug fixes**

- Make sure the dates in tracking fields are given on the UTC timezone


3.2.0 (2018-04-11)
------------------

**Deprecations**

- The collection specific settings must now be separated with ``.`` instead of ``_``.
  (eg. use ``kinto.signer.staging.certificates.editors_group`` instead of ``kinto.signer.staging_certificates.editors_group``) (fixes #224)

**New features**

- Give write permission to reviewers/editors groups on newly created collections (fixes #237)
- The preview collection signature is now refreshed along the destination (fixes #236)
- Tracking fields are now documented and new ones were added (``last_edit_date``, ``last_request_review_date``, ``last_review_date`` and ``last_signature_date``) (fixes #137)

**Internal changes**

- Now log an INFO message when the CloudFront invalidation request is sent (fixes #238)


3.1.0 (2018-03-16)
------------------

**New features**

- Cloudfront invalidation paths can be configured
- User does not have to be in the *reviewers* group to refresh a signature (fixes #233)

**Internal changes**

- Got rid of ``six`` since *kinto-signer* is Python 3 only.


3.0.0 (2018-03-08)

scripts/e2e.py
import random
from string import hexdigits
import argparse

from kinto_http import Client
from kinto_signer.serializer import canonical_json
from kinto_signer.hasher import compute_hash
from kinto_signer.signer.local_ecdsa import ECDSASigner


DEFAULT_SERVER = "http://localhost:8888/v1"
DEFAULT_AUTH = "user:pass"
DEFAULT_SOURCE_BUCKET = "alice"
DEFAULT_SOURCE_COL = "source"


def _rand(size=10):
    return "".join([random.choice(hexdigits) for _ in range(size)])


def upload_records(client, num):
    records = []
    for i in range(num):
        data = {"one": _rand(1000)}
        record = client.create_record(data=data)
        records.append(record["data"])
    return records


def _get_args():
    parser = argparse.ArgumentParser(description="End-to-end signing test")

    parser.add_argument("--auth", help="Basic Authentication", type=str, default=DEFAULT_AUTH)

    parser.add_argument(
        "--editor-auth", help="Basic Authentication for editor", type=str, default=None
    )

    parser.add_argument(
        "--reviewer-auth", help="Basic Authentication for reviewer", type=str, default=None
    )

    parser.add_argument("--server", help="Kinto Server", type=str, default=DEFAULT_SERVER)

    parser.add_argument(
        "--source-bucket", help="Source bucket", type=str, default=DEFAULT_SOURCE_BUCKET
    )

    parser.add_argument(
        "--source-col", help="Source collection", type=str, default=DEFAULT_SOURCE_COL
    )

    parser.add_argument("--reset", help="Reset collection data", type=bool, default=False)

    return parser.parse_args()


def main():
    args = _get_args()

    client = Client(
        server_url=args.server,
        auth=tuple(args.auth.split(":")),
        bucket=args.source_bucket,
        collection=args.source_col,
    )

    if args.editor_auth is None:
        args.editor_auth = args.auth

    if args.reviewer_auth is None:
        args.reviewer_auth = args.auth

    editor_client = Client(
        server_url=args.server,
        auth=tuple(args.editor_auth.split(":")),
        bucket=args.source_bucket,
        collection=args.source_col,
    )
    reviewer_client = Client(
        server_url=args.server,
        auth=tuple(args.reviewer_auth.split(":")),
        bucket=args.source_bucket,
        collection=args.source_col,
    )

    # 0. initialize source bucket/collection (if necessary)
    server_info = client.server_info()
    editor_id = editor_client.server_info()["user"]["id"]
    reviewer_id = reviewer_client.server_info()["user"]["id"]
    print("Server: {0}".format(args.server))
    print("Author: {user[id]}".format(**server_info))
    print("Editor: {0}".format(editor_id))
    print("Reviewer: {0}".format(reviewer_id))

    # 0. check that this collection is well configured.
    signer_capabilities = server_info["capabilities"]["signer"]

    resources = [
        r
        for r in signer_capabilities["resources"]
        if (args.source_bucket, args.source_col)
        == (r["source"]["bucket"], r["source"]["collection"])
        or (args.source_bucket, None) == (r["source"]["bucket"], r["source"]["collection"])
    ]
    assert len(resources) > 0, "Specified source not configured to be signed"
    resource = resources[0]
    if "preview" in resource:
        print(
            f"Signoff: {resource['source']['bucket']}/{resource['source']['collection']}"
            f" => {resource['preview']['bucket']}/{resource['preview']['collection']}"
            f" => {resource['destination']['bucket']}/{resource['destination']['collection']}"
        )
    else:
        print(
            f"Signoff: {resource['source']['bucket']}/{resource['source']['collection']} "
            f" => {resource['destination']['bucket']}/{resource['destination']['collection']}"
        )

    print("_" * 80)

    bucket = client.create_bucket(if_not_exists=True)
    client.create_collection(
        permissions={"write": [editor_id, reviewer_id] + bucket["permissions"]["write"]},
        if_not_exists=True,
    )

    editors_group = resource.get("editors_group") or signer_capabilities["editors_group"]
    editors_group = editors_group.format(collection_id=args.source_col)
    client.patch_group(id=editors_group, data={"members": [editor_id]})

    reviewers_group = resource.get("reviewers_group") or signer_capabilities["reviewers_group"]
    reviewers_group = reviewers_group.format(collection_id=args.source_col)
    client.patch_group(id=reviewers_group, data={"members": [reviewer_id]})

    if args.reset:
        client.delete_records()
        existing = 0
    else:
        existing_records = client.get_records()
        existing = len(existing_records)

    dest_col = resource["destination"].get("collection") or args.source_col
    dest_client = Client(
        server_url=args.server, bucket=resource["destination"]["bucket"], collection=dest_col
    )

    preview_client = None
    if "preview" in resource:
        preview_bucket = resource["preview"]["bucket"]
        preview_collection = resource["preview"].get("collection") or args.source_col
        preview_client = Client(
            server_url=args.server, bucket=preview_bucket, collection=preview_collection
        )

    # 1. upload data
    print("Author uploads 20 random records")
    records = upload_records(client, 20)

    # 2. ask for a signature
    # 2.1 ask for review (noop on old versions)
    print("Editor asks for review")
    data = {"status": "to-review"}
    editor_client.patch_collection(data=data)
    # 2.2 check the preview collection (if enabled)
    if preview_client:
        print("Check preview collection")
        preview_records = preview_client.get_records()
        expected = existing + 20
        assert len(preview_records) == expected, "%s != %s records" % (
            len(preview_records),
            expected,
        )
        metadata = preview_client.get_collection()["data"]
        preview_signature = metadata.get("signature")
        assert preview_signature, "Preview collection not signed"
        preview_timestamp = preview_client.get_records_timestamp()
    # 2.3 approve the review
    print("Reviewer approves and triggers signature")
    data = {"status": "to-sign"}
    reviewer_client.patch_collection(data=data)

    # 3. upload more data
    print("Author creates 20 others records")
    upload_records(client, 20)

    print("Editor updates 5 random records")
    for toupdate in random.sample(records, 5):
        editor_client.patch_record(data=dict(newkey=_rand(10), **toupdate))

    print("Author deletes 5 random records")
    for todelete in random.sample(records, 5):
        client.delete_record(id=todelete["id"])

    expected = existing + 20 + 20 - 5

    # 4. ask again for a signature
    # 2.1 ask for review (noop on old versions)
    print("Editor asks for review")
    data = {"status": "to-review"}
    editor_client.patch_collection(data=data)
    # 2.2 check the preview collection (if enabled)
    if preview_client:
        print("Check preview collection")
        preview_records = preview_client.get_records()
        assert len(preview_records) == expected, "%s != %s records" % (
            len(preview_records),
            expected,
        )
        # Diff size is 20 + 5 if updated records are also all deleted,
        # or 30 if deletions and updates apply to different records.
        diff_since_last = preview_client.get_records(_since=preview_timestamp)
        assert 25 <= len(diff_since_last) <= 30, "Changes since last signature are not consistent"

        metadata = preview_client.get_collection()["data"]
        assert preview_signature != metadata["signature"], "Preview collection not updated"

    # 2.3 approve the review
    print("Reviewer approves and triggers signature")
    data = {"status": "to-sign"}
    reviewer_client.patch_collection(data=data)

    # 5. wait for the result

    # 6. obtain the destination records and serialize canonically.

    records = list(dest_client.get_records())
    assert len(records) == expected, "%s != %s records" % (len(records), expected)
    timestamp = dest_client.get_records_timestamp()
    serialized = canonical_json(records, timestamp)
    print("Hash is %r" % compute_hash(serialized))

    # 7. get back the signed hash

    signature = dest_client.get_collection()["data"]["signature"]

    assert len(signature["public_key"]), "Signer returned empty public key"

    with open("pub", "w") as f:
        f.write(signature["public_key"])

    # 8. verify the signature matches the hash
    signer = ECDSASigner(public_key="pub")
    try:
        signer.verify(serialized, signature)
        print("Signature OK")
    except Exception:
        print("Signature KO")
        raise


if __name__ == "__main__":
    main()
